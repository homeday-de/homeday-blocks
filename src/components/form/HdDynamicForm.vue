<template>
  <form class="dynamicForm" @submit.prevent="submit" novalidate>
    <slot name="before"/>
    <div v-for="(line, i) in lines" class="dynamicForm__line" :key="`line-${i}`">
      <component
      v-for="item in getItemsArray(line)"
      ref="fields"
      :is="getComponent(item.type)"
      :key="`input-${item.name}`"
      v-model="formData[item.name]"
      v-bind="item.props"
      :name="item.name"
      class="dynamicForm__line__item"
      :lang="lang"
      />
    </div>
    <slot name="before-button"/>
    <HdButton
    v-if="submitLabel"
    modifier="primary"
    class="dynamicForm__submit"
    >
    {{ submitLabel }}
    </HdButton>
    <slot name="after"/>
  </form>
</template>

<script>
import HdInput from 'homeday-blocks/src/components/form/HdInput.vue';
import HdCheckbox from 'homeday-blocks/src/components/form/HdCheckbox.vue';
import HdRadio from 'homeday-blocks/src/components/form/HdRadio.vue';
import HdInputPassword from 'homeday-blocks/src/components/form/HdInputPassword.vue';
import HdPasswordConfirm from 'homeday-blocks/src/components/form/HdPasswordConfirm.vue';
import HdTextarea from 'homeday-blocks/src/components/form/HdTextarea.vue';
import HdSelect from 'homeday-blocks/src/components/form/HdSelect.vue';
import HdTagsSelector from 'homeday-blocks/src/components/form/HdTagsSelector.vue';
import HdSplitInput from 'homeday-blocks/src/components/form/HdSplitInput.vue';
import HdGoogleAutocomplete from 'homeday-blocks/src/components/form/HdGoogleAutocomplete.vue';
import HdButton from 'homeday-blocks/src/components/buttons/HdButton.vue';

export default {
  name: 'HdDynamicForm',
  components: {
    HdButton,
  },
  props: {
    submitLabel: {
      type: String,
      default: '',
    },
    items: {
      type: Array,
      default: () => [],
    },
    lang: {
      type: String,
      default: undefined,
    },
  },
  data() {
    return {
      formData: {},
    };
  },
  watch: {
    formData: {
      deep: true,
      handler(formData) {
        this.$emit('input', formData);
      },
    },
  },
  created() {
    this.initializeFormData(this.items);
  },
  computed: {
    lines() {
      return this.items;
    },
  },
  methods: {
    initializeFormData(array) {
      const flattenedArray = array.reduce((flatArray, item) => [
        ...flatArray,
        ...this.getItemsArray(item),
      ], []);
      flattenedArray.forEach((item) => {
        this.$set(this.formData, item.name, item.initialValue);
      });
    },
    getInvalidFields() {
      return this.$refs.fields.filter((field) => {
        if (field.validate) {
          return !field.validate();
        }
        return false;
      });
    },
    validate() {
      return this.getInvalidFields().length === 0;
    },
    submit() {
      const invalidFields = this.getInvalidFields();

      this.$emit('submit', {
        data: this.formData,
        isValid: invalidFields.length === 0,
        invalidFields,
      });
    },
    // Return an array with one item if the current line is an object and not an array
    getItemsArray(line) {
      if (Array.isArray(line)) {
        return line;
      }
      return [line];
    },
    getComponent(type) {
      switch (type) {
        case 'input':
          return HdInput;
        case 'inputPassword':
          return HdInputPassword;
        case 'checkbox':
          return HdCheckbox;
        case 'radio':
          return HdRadio;
        case 'passwordConfirm':
          return HdPasswordConfirm;
        case 'textarea':
          return HdTextarea;
        case 'select':
          return HdSelect;
        case 'tagsSelector':
          return HdTagsSelector;
        case 'splitInput':
          return HdSplitInput;
        case 'googleAutocomplete':
          return HdGoogleAutocomplete;
        default:
          return null;
      }
    },
  },
};
</script>

<style lang="scss" scoped>
@import 'homeday-blocks/src/styles/mixins.scss';

.dynamicForm {
  &__line {
    display: flex;
    &__item {
      flex: 1;
      margin-left: $inline-xs;
      margin-right: $inline-xs;
      &:first-of-type {
        margin-left: 0;
      }
      &:last-of-type {
        margin-right: 0;
      }
    }
  }
  &__submit {
    width: 100%;
  }
}
</style>
