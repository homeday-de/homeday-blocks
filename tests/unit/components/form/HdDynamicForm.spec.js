import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdDynamicForm from '@/components/form/HdDynamicForm.vue';
import LOGIN_FORM from '@/stories/mocks/forms/LOGIN';

const LINE_SELECTOR = '.dynamicForm__line';
const ITEM_SELECTOR = '.dynamicForm__line__item';
const SUBMIT_SELECTOR = '.dynamicForm__submit';

describe('HdDynamicForm', () => {
  const wrapperFactory = wrapperFactoryBuilder(HdDynamicForm,
    {
      propsData: {
        items: LOGIN_FORM,
        submitLabel: 'submit',
      },
    });

  let wrapper;
  beforeEach(() => {
    wrapper = wrapperFactory();
  });

  test('renders all the items', () => {
    expect(wrapper.findAll(LINE_SELECTOR).length).toBe(LOGIN_FORM.length);
  });

  test('does not render sumit button when `submitLabel` is empty', () => {
    wrapper = wrapperFactory({
      propsData: {
        submitLabel: '',
      },
    });

    expect(wrapper.find(SUBMIT_SELECTOR).exists()).toBe(false);
  });

  test('does not render sumit button when `submitLabel` is missing', () => {
    wrapper = wrapperFactory({
      propsData: {
        submitLabel: undefined,
      },
    });

    expect(wrapper.find(SUBMIT_SELECTOR).exists()).toBe(false);
  });

  test('renders nested components', () => {
    wrapper = wrapperFactory({
      propsData: {
        items: [
          [
            {
              type: 'input',
              name: 'input1',
            },
            {
              type: 'input',
              name: 'input2',
            },
          ],
        ],
      },
    });

    expect(wrapper.find(LINE_SELECTOR).findAll(ITEM_SELECTOR).length).toBe(2);
  });

  test('sets the initial values', () => {
    const formData = LOGIN_FORM.reduce((data, { name, initialValue }) => ({
      ...data,
      [name]: initialValue,
    }), {});

    expect(wrapper.vm.formData).toEqual(formData);
  });

  test('emits the right payload on input', () => {
    const TEST_VALUE = 'test value';
    const TEST_INPUT_NAME = LOGIN_FORM[0].name;
    wrapper.find(`[name=${TEST_INPUT_NAME}]`).setValue(TEST_VALUE);

    expect(wrapper.emitted('input')[0][0][TEST_INPUT_NAME]).toBe(TEST_VALUE);
  });

  describe('emits submit event', () => {
    test('invalid form', () => {
      wrapper.find('form').trigger('submit');
      const payload = wrapper.emitted('submit')[0][0];

      expect(payload.isValid).toBe(false);
      expect(payload.invalidFields.length).toBe(1);
    });

    test('valid form', () => {
      const TEST_VALUE = 'test value';
      const TEST_INPUT_NAME = LOGIN_FORM[1].name;
      wrapper.find(`[name=${TEST_INPUT_NAME}]`).setValue(TEST_VALUE);
      wrapper.find('form').trigger('submit');
      const payload = wrapper.emitted('submit')[0][0];

      expect(payload.isValid).toBe(true);
      expect(payload.invalidFields.length).toBe(0);
      expect(payload.data).toEqual({
        [LOGIN_FORM[0].name]: LOGIN_FORM[0].initialValue,
        [LOGIN_FORM[1].name]: TEST_VALUE,
      });
    });
  });

  test('renders slots', () => {
    wrapper = wrapperFactory({
      propsData: {
        items: [{}],
      },
      slots: {
        before: '<div>The before slot</div>',
        'before-button': '<div>The before button slot</div>',
        after: '<div>The after slot</div>',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
