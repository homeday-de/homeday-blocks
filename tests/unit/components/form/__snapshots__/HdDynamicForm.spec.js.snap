// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdDynamicForm renders slots 1`] = `
<form novalidate="novalidate" class="dynamicForm">
  <div>The before slot</div>
  <div class="dynamicForm__line">
    <!---->
  </div>
  <div>The before button slot</div> <button class="dynamicForm__submit btn btn--primary">
    submit
  </button>
  <div>The after slot</div>
</form>
`;
